File: lib/src/main/java/com/orientsec/easysocket/client/SocketSession.java
package com.orientsec.easysocket.client;

import android.net.TrafficStats;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.orientsec.easysocket.Address;
import com.orientsec.easysocket.EasySocket;
import com.orientsec.easysocket.Initializer;
import com.orientsec.easysocket.Options;
import com.orientsec.easysocket.Packet;
import com.orientsec.easysocket.PacketHandler;
import com.orientsec.easysocket.PacketType;
import com.orientsec.easysocket.Period;
import com.orientsec.easysocket.error.EasyException;
import com.orientsec.easysocket.error.ErrorBuilder;
import com.orientsec.easysocket.error.ErrorCode;
import com.orientsec.easysocket.error.ErrorType;
import com.orientsec.easysocket.task.TaskManager;
import com.orientsec.easysocket.utils.LogFactory;
import com.orientsec.easysocket.utils.Logger;

import java.io.IOException;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketAddress;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executor;

import javax.net.ssl.SSLSocket;

/**
 * Product: EasySocket
 * Package: com.orientsec.easysocket.inner.blocking
 * Time: 2017/12/27 14:35
 * Author: Fredric
 * coding is art not science
 */
public class SocketSession implements OperableSession, Initializer.Emitter, Runnable, EventListener {
    static final int ERROR = 202;
    static final int START_SUCCESS = 203;
    static final int START_FAILED = 204;
    static final int AVAILABLE = 205;
    static final int INIT_FAILED = 206;
    static final int ON_PACKET = 207;

    private Socket socket;

    private BlockingReader reader;

    private BlockingWriter writer;

    private final Executor connectExecutor;

    private final Options options;

    private final EventManager eventManager;

    private final Logger logger;

    private final AbstractSocketClient socketClient;

    private final Address address;

    private final int addressIndex;

    private State state = State.IDLE;

    private boolean serverAvailable = false;

    private Pulse pulse;

    private final Map<String, PacketHandler> messageHandlerMap = new HashMap<>();

    private final long id;

    private final ErrorBuilder errorBuilder;

    /**
     * 连接各阶段耗时。
     */
    private final Map<Period, Long> connectTimeMap = new HashMap<>();

    SocketSession(AbstractSocketClient socketClient, Address address, int addressIndex, long id) {
        this.socketClient = socketClient;
        this.address = address;
        this.addressIndex = addressIndex;
        this.options = socketClient.getOptions();
        this.id = id;

        connectExecutor = options.getConnectExecutor();

        eventManager = EasySocket.getInstance().newEventManager();
        eventManager.addListener(this);

        String suffix = "  Session(" + id + ")[" + address.getHost() + ":"
                + address.getPort() + "]  Client[" + options.getName() + "]";
        errorBuilder = new ErrorBuilder(suffix);
        logger = LogFactory.getLogger(options, suffix);
    }

    public Socket getSocket() {
        return socket;
    }

    @Override
    public Logger getLogger() {
        return logger;
    }

    @Override
    public void handlePacket(@NonNull Packet packet) {
        eventManager.publish(ON_PACKET, packet);
    }

    private void onPacket(@NonNull Packet packet) {
        if (state == State.DETACH) return;
        PacketHandler packetHandler = messageHandlerMap.get(packet.getPacketType().getValue());
        if (packetHandler == null) {
            logger.w("No packet handler for type: " + packet.getPacketType());
        } else {
            packetHandler.handlePacket(packet);
        }
    }


    @Override
    public void open() {
        if (state == State.IDLE) {
            connectExecutor.execute(this);
            state = State.STARTING;
            logger.i("Session is opening.");

            socketClient.onConnectionStart(this);
        }
    }

    @Override
    public void close(int code, int type, String message) {
        EasyException e = errorBuilder.create(code, type, message);
        if (state == State.IDLE || state == State.STARTING) {
            state = State.DETACH;
            logger.e("Session is closed.", e);
            socketClient.onConnectionFailed(this, e);
        } else {
            onError(e);
        }
    }

    @Override
    public void success() {
        eventManager.publish(AVAILABLE, this);
    }

    @Override
    public void fail(Exception cause) {
        logger.e("Fail to initialize session.");
        eventManager.publish(INIT_FAILED,
                errorBuilder.create(ErrorCode.SESSION_INIT_FAILED, ErrorType.CONNECT,
                        "Session initialize failed.", cause));
    }

    @Override
    public void postClose(int code, int type, String message, Exception cause) {
        eventManager.publish(ERROR, errorBuilder.create(code, type, message, cause));
    }

    private void onReady(Socket socket) {
        if (state == State.STARTING) {
            this.socket = socket;
            //连接后进行一些前置操作，例如资源初始化
            socketClient.getInitializer().start(this);
            //启动心跳及读写线程
            TaskManager taskManager = socketClient.getTaskManager();
            writer = new BlockingWriter(this, socket, taskManager);
            reader = new BlockingReader(this, socket, options, socketClient.getHeadParser());
            writer.start();
            reader.start();

            messageHandlerMap.put(PacketType.RESPONSE.getValue(), taskManager);

            state = State.CONNECT;
            logger.i("Session start success.");

            socketClient.onConnected(this);
        } else {
            //Session已关闭。
            connectExecutor.execute(() -> {
                try {
                    socket.close();
                } catch (IOException e) {
                    logger.w("Socket is closed.", e);
                }
            });
        }
    }

    private void onFailed(EasyException e) {
        if (state == State.STARTING) {
            state = State.DETACH;
            logger.e("Session start failed.");

            socketClient.onConnectionFailed(this, e);
        }
    }

    private void onAvailable() {
        if (state == State.CONNECT) {
            //开启心跳
            pulse = new Pulse(socketClient, this, eventManager);
            pulse.start();
            //注册主动心跳及推送消息处理器
            messageHandlerMap.put(PacketType.PULSE.getValue(), pulse);
            messageHandlerMap.put(PacketType.PUSH.getValue(), socketClient.getPushManager());

            state = State.AVAILABLE;
            serverAvailable = true;
            logger.i("Session is available.");

            socketClient.onConnectionAvailable(this);
        }
    }

    private void onError(EasyException e) {
        if (state == State.CONNECT || state == State.AVAILABLE) {
            if (pulse != null) {
                pulse.stop();
            }
            reader.shutdown();
            writer.shutdown();
            connectExecutor.execute(() -> {
                try {
                    socket.close();
                } catch (IOException ioe) {
                    logger.w("Socket is closed.", ioe);
                }
            });

            state = State.DETACH;
            logger.e("Session is closed.", e);

            socketClient.onDisconnected(this, e);
        }
    }

    /**
     * 启动socket连接。
     */
    @Override
    public void run() {
        logger.i("Socket connection is starting.");
        TrafficStats.setThreadStatsTag((int) Thread.currentThread().getId());
        try {
            Socket socket = socketClient.getSocketFactory().createSocket();
            //关闭Nagle算法,无论TCP数据报大小,立即发送
            socket.setTcpNoDelay(true);
            socket.setKeepAlive(true);
            socket.setPerformancePreferences(1, 2, 0);

            long startTimeMill = System.currentTimeMillis();
            long timestamp = startTimeMill;
            //STEP1:DNS
            SocketAddress socketAddress
                    = new InetSocketAddress(address.getHost(), address.getPort());
            long currentTimeMillis = System.currentTimeMillis();
            connectTimeMap.put(Period.DNS, currentTimeMillis - timestamp);
            timestamp = currentTimeMillis;

            //STEP2:CONNECT
            socket.connect(socketAddress, options.getConnectTimeOut());
            currentTimeMillis = System.currentTimeMillis();
            connectTimeMap.put(Period.CONNECT, currentTimeMillis - timestamp);
            timestamp = currentTimeMillis;

            //STEP3:SSL
            if (socket instanceof SSLSocket) {
                ((SSLSocket) socket).startHandshake();
                currentTimeMillis = System.currentTimeMillis();
                connectTimeMap.put(Period.SSL, currentTimeMillis - timestamp);
                timestamp = currentTimeMillis;
            }

            //总时长
            long connectTime = timestamp - startTimeMill;
            connectTimeMap.put(Period.ALL, connectTime);

            eventManager.publish(START_SUCCESS, socket);
            logger.i("Socket connected in " + connectTime + "ms");
        } catch (Exception e) {
            logger.e("Socket connection failed.", e);
            eventManager.publish(START_FAILED,
                    errorBuilder.create(ErrorCode.SOCKET_CONNECT, ErrorType.CONNECT,
                            "Socket connection failed.", e));
        }
    }

    @Override
    public boolean isConnect() {
        return state == State.CONNECT || state == State.AVAILABLE;
    }

    @Override
    public boolean isAvailable() {
        return state == State.AVAILABLE;
    }

    @Override
    public boolean isServerAvailable() {
        return serverAvailable;
    }

    @Override
    public void onEvent(int eventId, @Nullable Object object) {
        switch (eventId) {
            case ERROR:
                assert object != null;
                onError((EasyException) object);
                break;
            case AVAILABLE:
                onAvailable();
                break;
            case INIT_FAILED:
                onError((EasyException) object);
                break;
            case START_SUCCESS:
                assert object != null;
                onReady((Socket) object);
                break;
            case START_FAILED:
                assert object != null;
                onFailed((EasyException) object);
                break;
            case ON_PACKET:
                assert object != null;
                onPacket((Packet) object);
                break;
            case Pulse.PULSE:
                pulse.pulse();
                break;
        }
    }

    @NonNull
    @Override
    public Address getAddress() {
        return address;
    }

    @Nullable
    @Override
    public InetAddress getInetAddress() {
        if (socket != null) {
            return socket.getInetAddress();
        }
        return null;
    }

    @Override
    public int getAddressIndex() {
        return addressIndex;
    }

    @Override
    public long connectTime() {
        Long time = connectTimeMap.get(Period.ALL);
        if (time == null) {
            return -1;
        } else {
            return time;
        }
    }

    @Override
    public long connectTime(Period period) {
        Long time = connectTimeMap.get(period);
        if (time == null) {
            return -1;
        } else {
            return time;
        }
    }

    @NonNull
    @Override
    public String toString() {
        return "SocketSession[" +
                "id=" + id +
                "address=" + address +
                ']';
    }
}

File: lib-socket/src/androidMain/kotlin/com/orientsec/easysocket/socket/SocketSession.kt
package com.orientsec.easysocket.socket

import com.orientsec.easysocket.Address
import com.orientsec.easysocket.Period
import com.orientsec.easysocket.client.BaseSocketClient
import com.orientsec.easysocket.error.EasyException
import com.orientsec.easysocket.error.ErrorCode
import com.orientsec.easysocket.error.ErrorType
import com.orientsec.easysocket.session.AbstractSession
import com.orientsec.easysocket.session.Reader
import com.orientsec.easysocket.session.Writer
import com.orientsec.easysocket.utils.Platform
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import java.net.InetSocketAddress
import java.net.Socket
import java.net.SocketAddress
import javax.net.ssl.SSLSocket
import kotlin.system.measureTimeMillis
import kotlin.time.measureTimedValue

/**
 * 基于传统 Java Socket 的 Session 实现。
 *
 * 使用 java.net.Socket 进行连接建立和数据读写。
 * 支持 TCP 连接和 SSL/TLS 加密连接（通过 SSLSocket）。
 * 连接过程会记录各阶段（DNS、CONNECT、SSL）的耗时，
 * 并使用 TrafficProfiler 标记 Socket 流量。
 *
 * @param socketClient 所属的 Socket 客户端
 * @param address 连接的服务器地址
 * @param addressIndex 地址在列表中的索引
 * @param id 会话唯一标识
 * @param socketFactory Socket 工厂，用于创建 Socket 实例，默认使用系统默认工厂
 */
class SocketSession(
    socketClient: BaseSocketClient,
    address: Address,
    addressIndex: Int,
    id: Long,
    private val socketFactory: javax.net.SocketFactory = javax.net.SocketFactory.getDefault()
) : AbstractSession(socketClient, address, addressIndex, id) {

    /** 底层 Java Socket 实例 */
    private var mSocket: Socket? = null

    /**
     * 执行具体的连接逻辑。
     * 使用传统 Java Socket API 建立 TCP 连接，如果需要则进行 SSL 握手。
     * 连接过程中会记录各阶段（DNS、CONNECT、SSL）的耗时。
     *
     * @return true 如果连接成功，false 如果连接失败
     */
    override suspend fun performConnect(): Boolean {
        logger.d("socket connection is starting")
        val startTimeInMills = Platform.currentTimeMillis()
        val connectResult = withContext(Dispatchers.IO) {
            // 1. DNS 解析
            val dnsResult = dns()
            if (dnsResult is ConnectResult.Failure) return@withContext dnsResult
            dnsResult as ConnectResult.Success
            connectTimeMap[Period.DNS] = dnsResult.timeInMills
            val socketAddress: SocketAddress = dnsResult.value

            // 2. TCP 连接阶段
            val socketResult = connect(socketAddress)
            if (socketResult is ConnectResult.Failure) return@withContext socketResult
            socketResult as ConnectResult.Success
            connectTimeMap[Period.CONNECT] = socketResult.timeInMills
            val socket = socketResult.value

            // 3. SSL 握手阶段
            if (address.isSsl) {
                ssl(socket).also {
                    if (it is ConnectResult.Success) {
                        connectTimeMap[Period.SSL] = it.timeInMills
                    }
                }
            } else {
                socketResult
            }
        }

        return when (connectResult) {
            is ConnectResult.Failure -> {
                logger.w("socket connection failed", connectResult.error)
                EasyException(
                    connectResult.errorCode,
                    ErrorType.CONNECT,
                    "socket connection failed",
                    suffix,
                    connectResult.error
                ).let { onFailed(it) }
                false
            }

            is ConnectResult.Success -> {
                logger.d(
                    "socket connected in " +
                            "${Platform.currentTimeMillis() - startTimeInMills}ms"
                )
                this.mSocket = connectResult.value
                true
            }
        }
    }

    private fun dns(): ConnectResult<SocketAddress> {
        // 1. DNS 解析
        return try {
            measureTimedValue {
                InetSocketAddress(address.host, address.port)
            }.let {
                ConnectResult.success(
                    value = it.value,
                    timeInMills = it.duration.inWholeMilliseconds
                )
            }
        } catch (e: Exception) {
            ConnectResult.failure(
                error = e,
                errorCode = ErrorCode.DNS_ANALYZE
            )
        }
    }

    // 2. TCP 连接阶段
    private fun connect(socketAddress: SocketAddress): ConnectResult<Socket> {
        val socket = try {
            socketFactory.createSocket()
        } catch (e: Exception) {
            return ConnectResult.failure(
                error = e,
                errorCode = ErrorCode.SOCKET_CREATE
            )
        }
        options.trafficProfiler.tagSocket(socket)
        socket.tcpNoDelay = true
        socket.keepAlive = true
        socket.setPerformancePreferences(1, 2, 0)
        return try {
            val timeInMills = measureTimeMillis {
                socket.connect(socketAddress, options.connectTimeoutMills)
            }
            ConnectResult.success(socket, timeInMills)
        } catch (e: Exception) {
            options.trafficProfiler.untagSocket(socket)
            try {
                socket.close()
            } catch (_: Exception) {
            }
            ConnectResult.failure(
                error = e,
                errorCode = ErrorCode.SOCKET_CONNECT
            )
        }
    }

    private fun ssl(socket: Socket): ConnectResult<Socket> {
        if (socket !is SSLSocket) {
            options.trafficProfiler.untagSocket(socket)
            try {
                socket.close()
            } catch (_: Exception) {
            }
            return ConnectResult.failure(
                error = IllegalStateException("Not an SSLSocket"),
                errorCode = ErrorCode.TLS_ERROR
            )
        }

        return try {
            socket.soTimeout = options.tlsTimeoutMills
            val timeInMills = measureTimeMillis {
                socket.startHandshake()
            }
            socket.soTimeout = 0
            ConnectResult.success(value = socket, timeInMills = timeInMills)
        } catch (e: Exception) {
            options.trafficProfiler.untagSocket(socket)
            try {
                socket.close()
            } catch (_: Exception) {
            }
            val errorCode = if (e is java.net.SocketTimeoutException) ErrorCode.TLS_TIMEOUT
            else ErrorCode.TLS_ERROR
            ConnectResult.failure(error = e, errorCode = errorCode)
        }
    }

    /**
     * 连接结果封装类。
     */
    private sealed class ConnectResult<out T> {
        class Success<T>(
            val value: T,
            val timeInMills: Long
        ) : ConnectResult<T>()


        class Failure(
            val error: Exception,
            val errorCode: Int,
        ) : ConnectResult<Nothing>()

        companion object {
            fun <T> success(
                value: T,
                timeInMills: Long
            ): Success<T> {
                return Success(value, timeInMills)
            }

            fun failure(error: Exception, errorCode: Int): Failure {
                return Failure(error, errorCode)
            }
        }
    }

    /**
     * 关闭底层 Java Socket。
     * 取消流量标记后在 IO 线程中关闭 Socket。
     */
    override fun closeSocket() {
        val socket = mSocket
        if (socket != null) {
            options.trafficProfiler.untagSocket(socket)
            socketClient.scope.launch(Dispatchers.IO) {
                try {
                    socket.close()
                } catch (_: Exception) {
                }
            }
        }
    }

    /**
     * 创建阻塞式读取器。
     *
     * @return [BlockingReader] 实例
     */
    override fun createReader(): Reader {
        return BlockingReader(this, socketClient, mSocket!!)
    }

    /**
     * 创建队列式写入器。
     *
     * @return [QueuedWriter] 实例
     */
    override fun createWriter(): Writer {
        return QueuedWriter(this, socketClient.scope, mSocket!!)
    }

    /**
     * 获取连接的 IP 地址。
     *
     * @return IP 地址字符串，未连接时返回 null
     */
    override val ipAddress: String? get() = mSocket?.inetAddress?.hostAddress

}